akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardingSerializer.scala

```diff
@@ -52,4 +56,28 @@ import akka.serialization.SerializerWithStringManifest
         s"Unimplemented deserialization of message with manifest [$manifest] in [${getClass.getName}]")
   }
 
+  // buffer based avoiding a copy for artery
+  override def toBinary(o: AnyRef, buf: ByteBuffer): Unit = o match {
+    case env: ShardingEnvelope[_] =>
+      val builder = ShardingMessages.ShardingEnvelope.newBuilder()
+      builder.setEntityId(env.entityId)
+      builder.setMessage(payloadSupport.payloadBuilder(env.message))
```

Comment:
do you think there is a way to use the `buf` all the way into `WrappedPayloadSupport` so that if the actual serializer for the message is also a ByteBufferSerializer we could write directly without creating array?

Reply:
I thought about it, but perhaps not enough, I'll see if I can figure out a way.

Reply:
Mentioned elsewhere but for the record:

We can't benefit from passing the byte buffer on to WrappedPayloadSupport since protobuf encodes the size of a nested message as a varint, and we cannot know the size of the payload serialized with an arbitrary serializer up front.

An envelope serializer in a specific application, knowing that the nested payload is protobuf could potentially benefit from such an optimization (getting calculating the size before serializing the nested data).